File: bindings/python/perf/multi/perf_multi_dealer_dealer_client.py
import asyncio
import sys
import time
from contextlib import ExitStack

import zlink

from perf_multi_common import (
    STOP_TOKEN,
    apply_multi_socket_options,
    benchmark_run_id,
    configure_multi_tls_client,
    MultiSendTurnCoordinator,
    new_payload,
    parse_client_args,
    perf_client_context,
    print_multi_auto_hwm_detail,
    resolve_multi_monitor_hwm_bytes,
    resolve_multi_connect_ready_timeout_ms,
    resolve_multi_send_drain_timeout_ms,
    send_routed,
    stamp_payload,
    submit_routed,
    wait_monitor_event,
)


async def _send_stop_token(sock, completion_poller, completion_events):
    await send_routed(
        sock,
        STOP_TOKEN,
        completion_poller=completion_poller,
        completion_events=completion_events,
        measurement=False,
    )


async def main(argv=None):
    args = parse_client_args(argv or sys.argv[1:], pattern="dealer_dealer")
    run_id = benchmark_run_id()
    payloads = [new_payload(args.msg_size) for _ in range(args.clients)]
    seq = 0

    with perf_client_context() as ctx:
        sockets = [zlink.create_dealer_socket(ctx) for _ in range(args.clients)]
        try:
            with ExitStack() as stack:
                monitors = []
                for sock in sockets:
                    monitor = stack.enter_context(
                        sock.monitor_open(
                            zlink.MonitorEventMask.CONNECTION_READY,
                            resolve_multi_monitor_hwm_bytes(),
                        )
                    )
                    configure_multi_tls_client(sock, args.transport)
                    apply_multi_socket_options(sock)
                    sock.connect(args.endpoint)
                    monitors.append(monitor)
                for monitor in monitors:
                    wait_monitor_event(
                        monitor,
                        zlink.MonitorEventMask.CONNECTION_READY,
                        timeout_ms=resolve_multi_connect_ready_timeout_ms(),
                    )
                # Readiness ownership ends at the barrier. Release these
                # monitors before the measured-path HWM snapshot opens one.
                stack.close()
                print(f"CLIENT_READY,{args.msg_size}", flush=True)
                command = sys.stdin.readline().strip()
                if command != f"START,{args.msg_size}":
                    raise SystemExit(f"unexpected command: {command}")

                active_deadline = time.perf_counter() + args.duration
                with zlink.create_poller() as completion_poller:
                    completion_events = zlink.create_poll_events(
                        max(1, len(sockets))
                    )
                    for index, sock in enumerate(sockets):
                        completion_poller.add_socket(
                            sock,
                            zlink.PollEventFlag.POLLCOMPLETION,
                            index,
                        )

                    def submit(index):
                        nonlocal seq
                        seq += 1
                        return submit_routed(
                            sockets[index],
                            stamp_payload(
                                payloads[index],
                                phase=1,
                                run_id=run_id,
                                seq=seq,
                            ),
                        )

                    coordinator = MultiSendTurnCoordinator(
                        completion_poller,
                        completion_events,
                        len(sockets),
                    )
                    await coordinator.run(
                        active_deadline,
                        submit,
                        drain_timeout_ms=resolve_multi_send_drain_timeout_ms(),
                    )
                    for current_sock in sockets:
                        await _send_stop_token(
                            current_sock,
                            completion_poller,
                            completion_events,
                        )
                if sockets:
                    print_multi_auto_hwm_detail(
                        sockets[0], "endpoint", args.transport, args.msg_size, "dealer"
                    )
                # C run_single_size_case: send a wire stop token per socket
                # so the server receive window terminates.
                print(f"CLIENT_DONE,{args.msg_size}", flush=True)
        finally:
            for sock in sockets:
                try:
                    sock.close()
                except Exception as exc:
                    print(f"[perf] close failed: {exc}", file=sys.stderr)


if __name__ == "__main__":
    asyncio.run(main())

File: bindings/node/perf/multi/perf_multi_dealer_dealer_client.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const readline = require('node:readline');
const zlink = require('@zlink-systems/zlink');
const {
  currentEpochNs,
  createPayload,
  stampPayload,
} = require('../common/perf_metrics');
const { configureTlsClient } = require('../common/perf_tls');
const { parseMultiArgs } = require('./perf_multi_common');
const {
  applyContextPolicy,
  applySocketPolicy,
  emitMultiSocketHwmDetail,
  sendRouted,
  waitForConnectionReady
} = require('./perf_multi_runtime');
const { STOP_TOKEN_BYTES } = require('../perf_stop_token');

// MULTI_DEALER_DEALER client == SENDER (one DEALER socket per client).
//
// C parity: bindings/c/perf/multi/src/perf_multi_dealer_dealer_client.cpp
// is the SENDER. It creates one DEALER socket per client (connect),
// prints CLIENT_READY,<size>, waits START,<size> from stdin, runs the
// per-socket bounded send window (run_send_window ~142-265). This binding
// awaits the Core send-completion terminal instead of using the removed DONTWAIT terminal,
// then sends exactly ONE wire stop token per socket
// (run_single_size_case ~290-293 / send_stop_token ~114-140). The
// matching RECEIVER/MEASURER is perf_multi_dealer_dealer_server.cpp.
// Cross-checked against the already-fixed cpp
// bindings/cpp/perf/multi/src/perf_dealer_dealer_client.cpp.
// Handshake (PERF_MULTI § 1.5 / line 201): server READY,<endpoint> then
// client spawn; client prints CLIENT_READY,<size>; runner sends
// START,<size> to BOTH; sender runs the send window after START.
async function main() {
  const options = parseMultiArgs(process.argv.slice(2));
  const ctx = zlink.createContext();
  applyContextPolicy(ctx, 'client', 'MULTI_DEALER_DEALER');
  const dealers = [];
  let rl = null;

  try {
    for (let i = 0; i < options.clients; i += 1) {
      const dealer = zlink.createDealerSocket(ctx);
      applySocketPolicy(dealer, { transport: options.transport });
      configureTlsClient(dealer, options.transport);
      dealers.push(dealer);
    }
    for (let i = 0; i < dealers.length; i += 1) {
      const dealer = dealers[i];
      await waitForConnectionReady(dealer, () => dealer.connect(options.endpoint));
    }
    ctx.recalculateAutoHwm();
    for (const dealer of dealers) {
      emitMultiSocketHwmDetail(dealer, 'endpoint', options.transport, options.msgSize);
    }

    console.log(`CLIENT_READY,${options.msgSize}`);
    rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
    for await (const line of rl) {
      if (line === `START,${options.msgSize}`) {
        break;
      }
      if (line === 'STOP' || line === 'QUIT') {
        return;
      }
    }

    const payloads = dealers.map(() => createPayload(options.msgSize));
    const activeStopNs = currentEpochNs() + BigInt(Math.floor(options.duration * 1_000_000_000));
    let seq = 1n;
    await Promise.all(dealers.map(async (dealer, index) => {
      while (currentEpochNs() < activeStopNs) {
        const currentSeq = seq;
        seq += 1n;
        stampPayload(payloads[index], {
          phase: 1, runId: 1, msgSize: options.msgSize, seq: currentSeq
        });
        await sendRouted(dealer, payloads[index]);
      }
      await sendRouted(dealer, [STOP_TOKEN_BYTES]);
    }));
    console.log(`CLIENT_DONE,${options.msgSize}`);
  } finally {
    rl?.close();
    for (const dealer of dealers) {
      dealer.close();
    }
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
